File: common/src/main/java/gg/grounds/discovery/PaperServerEntry.java
package gg.grounds.discovery;

import java.util.Optional;

public record PaperServerEntry(String name, String host, int port) {

    public String encode() {
        return name + "|" + host + "|" + port;
    }

    public static Optional<PaperServerEntry> decode(String value) {
        if (value == null || value.isBlank()) {
            return Optional.empty();
        }

        String[] parts = value.split("\\|", -1);
        if (parts.length != 3) {
            return Optional.empty();
        }

        String name = parts[0].trim();
        String host = parts[1].trim();
        if (name.isEmpty() || host.isEmpty()) {
            return Optional.empty();
        }

        int port;
        try {
            port = Integer.parseInt(parts[2].trim());
        } catch (NumberFormatException e) {
            return Optional.empty();
        }

        try {
            return Optional.of(new PaperServerEntry(name, host, port));
        } catch (IllegalArgumentException e) {
            return Optional.empty();
        }
    }
}

File: common/src/main/kotlin/gg/grounds/discovery/PaperServerEntry.kt
package gg.grounds.discovery

data class PaperServerEntry(
    val name: String,
    val host: String,
    val port: Int,
    val lastSeenMillis: Long = 0L,
) {

    fun encode(): String = "$name|$host|$port|$lastSeenMillis"

    companion object {
        @JvmStatic
        fun decode(value: String?): PaperServerEntry? {
            if (value.isNullOrBlank()) {
                return null
            }

            val parts = value.split('|')
            if (parts.size != 4) {
                return null
            }

            val name = parts[0].trim().takeIf { it.isNotEmpty() } ?: return null
            val host = parts[1].trim().takeIf { it.isNotEmpty() } ?: return null
            val port = parts[2].trim().toIntOrNull() ?: return null
            val lastSeenMillis = parts[3].trim().toLongOrNull() ?: return null

            return PaperServerEntry(name, host, port, lastSeenMillis)
        }
    }
}
